File: android/control-proxy/src/main/kotlin/dev/jasonpearson/automobile/ctrlproxy/models/UIElementInfo.kt
package dev.jasonpearson.automobile.ctrlproxy.models

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.Transient
import kotlinx.serialization.json.JsonElement

/**
 * Data class representing UI elements with all relevant properties extracted from
 * AccessibilityNodeInfo for automated testing.
 *
 * Property names match the XML attribute format used by uiautomator to maintain compatibility with
 * existing test frameworks.
 */
@Serializable
data class UIElementInfo(
  val text: String? = null,
  val textSize: Float? = null,
  @SerialName("text-color") val textColor: String? = null, // Hex color string like "#FF000000"
  @SerialName("content-desc") val contentDesc: String? = null,
  @SerialName("resource-id") val resourceId: String? = null,
  @SerialName("view-id") val viewId: String? = null,
  val className: String? = null,
  val bounds: ElementBounds? = null,
  val clickable: String? = null, // "true"/"false" to match XML format
  val enabled: String? = null,
  val focusable: String? = null,
  val focused: String? = null,
  @SerialName("accessibility-focused") val accessibilityFocused: String? = null,
  val scrollable: String? = null,
  val password: String? = null,
  val checkable: String? = null,
  val checked: String? = null,
  val selected: String? = null,
  @SerialName("long-clickable") val longClickable: String? = null,
  val fragment: String? = null, // Fragment class name when applicable

  // Additional accessibility semantics fields
  @SerialName("test-tag") val testTag: String? = null, // Compose or View accessibility-extra tag
  @SerialName("unique-id") val uniqueId: String? = null, // Android-owned ID (API 33+)
  @SerialName("visible-to-user") val visibleToUser: Boolean? = null,
  @SerialName("container-title") val containerTitle: String? = null, // API 34+
  val role: String? = null, // Accessibility role (button, checkbox, etc.)
  @SerialName("state-description") val stateDescription: String? = null, // Custom state description
  @SerialName("error-message") val errorMessage: String? = null, // Error message for form fields
  @SerialName("hint-text") val hintText: String? = null, // Hint text for input fields
  @SerialName("tooltip-text") val tooltipText: String? = null, // Tooltip text
  @SerialName("pane-title") val paneTitle: String? = null, // Pane title for navigation
  @SerialName("live-region") val liveRegion: String? = null, // Live region mode
  @SerialName("collection-info") val collectionInfo: String? = null, // Collection information
  @SerialName("collection-item-info")
  val collectionItemInfo: String? = null, // Collection item info
  @SerialName("collection-row-index") val collectionRowIndex: Int? = null,
  @SerialName("collection-column-index") val collectionColumnIndex: Int? = null,
  @SerialName("range-info") val rangeInfo: String? = null, // Range information for sliders/progress
  @SerialName("input-type") val inputType: String? = null, // Input type for text fields
  @SerialName("actions") val actions: List<String>? = null, // Available accessibility actions
  @SerialName("extras") val extras: Map<String, String>? = null, // Custom extras from semantics
  val occlusionState: String? = null, // visible | partial | hidden
  val occludedBy: String? = null, // Resource ID or label of the occluding view
  @SerialName("occludedByViewId")
  val occludedByViewId: String? = null, // Stable view-id of the occluding view
  val recomposition: RecompositionEntry? = null,

  // Wire projection of the child subtree (issue #5471). This is the ONLY serialized
  // representation of children: `null` (leaf), a JSON object (single child), or a JSON array
  // (multiple children), matching the format existing desktop/MCP/TS consumers parse.
  //
  // During extraction this stays null — the pipeline (optimize, detectors, occlusion, focus
  // search, hashing) walks the typed [children] list below with zero (de)serialization. `node`
  // is materialized from [children] exactly once, at the wire boundary (see WireNodeCodec).
  val node: JsonElement? = null,

  // Typed, in-memory child list. Not serialized (see [node] for the wire form). Every intermediate
  // extraction pass operates on this list so a snapshot never re-encodes/re-decodes subtrees.
  @Transient val children: List<UIElementInfo> = emptyList(),
) {
  /** Helper properties for boolean checks (for backwards compatibility) */
  val isClickable: Boolean
    get() = clickable == "true"

  val isEnabled: Boolean
    get() = enabled != "false" // Default true if not specified

  val isFocusable: Boolean
    get() = focusable == "true"

  val isFocused: Boolean
    get() = focused == "true"

  val isAccessibilityFocused: Boolean
    get() = accessibilityFocused == "true"

  val isScrollable: Boolean
    get() = scrollable == "true"

  val isPassword: Boolean
    get() = password == "true"

  val isCheckable: Boolean
    get() = checkable == "true"

  val isChecked: Boolean
    get() = checked == "true"

  val isSelected: Boolean
    get() = selected == "true"

  val isLongClickable: Boolean
    get() = longClickable == "true"
}

File: ios/control-proxy/Sources/CtrlProxyRewrite/Models/UIElementInfo.swift
import Foundation

/// UI Element information (matching Android's UIElementInfo).
///
/// The custom `CodingKeys` map camelCase Swift properties to the kebab-case JSON
/// keys the TS client reads (`content-desc`, `resource-id`, `semantic-links`,
/// `accessibility-focused`, `long-clickable`, `view-id`, `state-description`,
/// `error-message`, `hint-text`). These names are the frozen wire contract and are
/// reproduced verbatim from the reference target.
public struct UIElementInfo: Codable, Sendable {
    public let text: String?
    public let value: String?
    public let textSize: Float?
    public let contentDesc: String?
    public let resourceId: String?
    public let className: String?
    public let bounds: ElementBounds?
    public let clickable: String?
    public let enabled: String?
    public let focusable: String?
    public let focused: String?
    public let accessibilityFocused: String?
    public let scrollable: String?
    public let password: String?
    public let checkable: String?
    public let checked: String?
    public let selected: String?
    public let longClickable: String?
    public let semanticLinks: [SemanticLink]?
    public let testTag: String?
    public let role: String?
    public let stateDescription: String?
    public let errorMessage: String?
    public let hintText: String?
    public let viewId: String?
    public let extras: [String: String]?
    public let actions: [String]?
    public let node: [UIElementInfo]?

    enum CodingKeys: String, CodingKey {
        case text, value, textSize, className, bounds, clickable, enabled
        case focusable, focused, scrollable, password, checkable, checked
        case selected, actions, node, role, testTag, extras
        case semanticLinks = "semantic-links"
        case viewId = "view-id"
        case contentDesc = "content-desc"
        case resourceId = "resource-id"
        case accessibilityFocused = "accessibility-focused"
        case longClickable = "long-clickable"
        case stateDescription = "state-description"
        case errorMessage = "error-message"
        case hintText = "hint-text"
    }

    public init(
        text: String? = nil,
        value: String? = nil,
        textSize: Float? = nil,
        contentDesc: String? = nil,
        resourceId: String? = nil,
        className: String? = nil,
        bounds: ElementBounds? = nil,
        clickable: String? = nil,
        enabled: String? = nil,
        focusable: String? = nil,
        focused: String? = nil,
        accessibilityFocused: String? = nil,
        scrollable: String? = nil,
        password: String? = nil,
        checkable: String? = nil,
        checked: String? = nil,
        selected: String? = nil,
        longClickable: String? = nil,
        semanticLinks: [SemanticLink]? = nil,
        testTag: String? = nil,
        role: String? = nil,
        stateDescription: String? = nil,
        errorMessage: String? = nil,
        hintText: String? = nil,
        viewId: String? = nil,
        extras: [String: String]? = nil,
        actions: [String]? = nil,
        node: [UIElementInfo]? = nil
    ) {
        self.text = text
        self.value = value
        self.textSize = textSize
        self.contentDesc = contentDesc
        self.resourceId = resourceId
        self.className = className
        self.bounds = bounds
        self.clickable = clickable
        self.enabled = enabled
        self.focusable = focusable
        self.focused = focused
        self.accessibilityFocused = accessibilityFocused
        self.scrollable = scrollable
        self.password = password
        self.checkable = checkable
        self.checked = checked
        self.selected = selected
        self.longClickable = longClickable
        self.semanticLinks = semanticLinks
        self.testTag = testTag
        self.role = role
        self.stateDescription = stateDescription
        self.errorMessage = errorMessage
        self.hintText = hintText
        self.viewId = viewId
        self.extras = extras
        self.actions = actions
        self.node = node
    }
}
